streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -107,6 +107,9 @@
     private boolean commitNeeded = false;
     private boolean commitRequested = false;
 
+    private boolean checkpointNeeded = false;
```

Comment:
Do we need two parameters here? Could we still just use one `Map<TopicPartition, Long> checkpoint`, and then in `writeCheckpointIfNeed` if it is not null write it and then set it to null.

Reply:
I actually had a similar thought, but I am torn though. Using two variables required to keep them "in sync" was is not great. However, using `null` is less explicit... Thus overall I am fine either way as both seems to provide the overall same good/bad ratio.

Reply:
A third option is Optional<Map<TopicPartition, Long>> which explicitly states the checkpoint  is null. However, I think the benefit here is minimal, as `checkpointNeeded` sounds like a better state variable, than relying on the struct itself.

Reply:
Actually I'm not too concerned of relying on `null` to indicate no need to checkpoint, what I originally pointed out is that we are unnecessarily accumulating and then distributing checkpoints between task manager and task, which is already resolved in your PR. So I'd suggest we still just use a single nullable `Map`.

Reply:
Cool